utils/src/main/java/software/amazon/awssdk/utils/async/FlatteningSubscriber.java

```diff
@@ -48,7 +48,9 @@ public void onSubscribe(Subscription subscription) {
             public void request(long l) {
                 synchronized (lock) {
                     demand.addAndGet(l);
-                    if (currentBatch.isEmpty() && !requestedNextBatch) {
+                    // Execution goes into `if` block only once for the initial request
```

Comment:
Not sure I fully follow here, why do we no longer check if currentBatch.isEmpty()

Reply:
1) When the request is called for first time, `requestedNextBatch` is false and `currentBatch` is empty. This is the only time when the code execution goes into the `if` block. As `currentBatch` is always empty for first request, there is no explicit check is needed
2) After first time, requestedNextBatch is always true and so the condition `!requestedNextBatch` will always be evaluated to `false`. So there is no need to again check for `currentBatch.isEmpty()` as its evaluation no longer matter because other condition always evaluates to false.